<img src="images/logodwengo.png" alt="Banner" width="150"/>

<div>
    <font color=#690027 markdown="1">
        <h1>ENKELE BEWERKINGEN BIJ NUMPY-LIJSTEN</h1> 
    </font>
</div>

<div class="alert alert-block alert-success"> 
In deze notebook krijg je voorbeelden voorgeschoteld van bewerkingen die je kan uitvoeren op NumPy-lijsten. 
</div>

Dit kwam reeds aan bod in de notebook 'Spreidingsdiagram - Data manueel ingeven' en wordt hier hernomen. 

### Module importeren

In [1]:
import numpy as np

<div>
    <font color=#690027 markdown="1">
        <h2>1. Data</h2> 
    </font>
</div>

Je zal met de vier NumPy-lijsten werken uit de volgende code-cel. Voer na de code-cel voor de import ook deze code-cel uit.  

In [2]:
x = np.array([28,  40,  41,  44,  45,  51,  54,  57,  57,  60,  62,  72,  74, 90, 113])
y = np.array([33,  35,  47,  35,  55,  56,  50,  50,  70,  65,  53,  82,  80, 95, 100])
rij = np.array([2, 7, 10, 5])
rij2 = np.array([2, 4, 6, 8, 10, 12, 14, 16, 18, 20])

<div>
    <font color=#690027 markdown="1">
        <h2>2. Enkele bewerkingen</h2> 
    </font>
</div>

Soms is het interessant om te weten waar een bepaalde waarde in een NumPy-lijst voorkomt, of wat het gemiddelde is van een lijst getallen. 

### Voorbeeld 2.1: index
Elk element in een NumPy-lijst heeft een **index**. Je ziet die vooraan in de tabel hierboven staan.
M.b.v. de index kan je een bepaald element van een NumPy-lijst opvragen. Zo is het eerste element in de NumPy-lijst `x` het element met index 0. Hoe je die waarde opvraagt, wordt getoond in de volgende code-cel: 

In [3]:
# element opvragen
x[0]                                            # element uit lijst x met index 0

28

In [4]:
# element opvragen
x[6]                                            # zevende element uit lijst x

54

### Opdracht 2.1
Vraag het negende element op van NumPy-lijst `y`.

In [5]:
# voorbeeldoplossing
y[8]

70

### Voorbeeld 2.2: gemiddelde en som van de elementen van een NumPy-lijst
Met de functie `mean()` uit de module NumPy kan je het gemiddelde berekenen en met de functie `sum()` de som. 

In [6]:
# gemiddelde berekenen
x_gemiddelde = np.mean(x)                         # gemiddelde van alle elementen van lijst x
y_gemiddelde_middelstevijf = np.mean(y[5:10])     # gemiddelde van y[5] t.e.m. y[9]
print(x_gemiddelde, y_gemiddelde_middelstevijf)   # laat beide resultaten zien, gescheiden door komma 

59.2 58.2


In [7]:
# som berekenen
x_som = np.sum(x)
print(x_som)

888


### Opdracht 2.2
- Bereken het gemiddelde van de elementen van de NumPy-lijst `y`.
- Bereken het gemiddelde van de eerste 10 elementen van de NumPy-lijst `x`. 
- Bereken de som van de elementen van de NumPy-lijst `y`.

In [8]:
# voorbeeldoplossing
y_gemiddelde = np.mean(y) 
x_gemiddelde_eerstetien = np.mean(x[0:10]) 
y_som = np.sum(y)

print(y_gemiddelde, y_som, x_gemiddelde_eerstetien)

60.4 906 47.7


Antwoord:

Antwoord:
-  het gemiddelde van de elementen van de NumPy-lijst `y` is 60,4;
-  het gemiddelde van de eerste 10 elementen van de NumPy-lijst `x` is 906
-  de som van de elementen van de NumPy-lijst `y` is 47,7.

### Voorbeeld 2.3: een element opzoeken
Met de functie `where()` van de module NumPy kan je achterhalen wat de index is van een bepaald element. 

In [9]:
# index van bepaalde waarde
np.where(x == 60)

(array([9], dtype=int64),)

In [10]:
# index van bepaalde waarde
np.where(x == 57)

(array([7, 8], dtype=int64),)

### Opdracht 2.3
- Het hoeveelste element van de NumPy-lijst `y` is 56?
- Welk element van de NumPy-lijst `x` heeft dezelfde index als 56 uit NumPy-lijst `y`?

In [11]:
# voorbeeldoplossing
np.where(y == 56)

(array([5], dtype=int64),)

In [12]:
# voorbeeldoplossing
x[5]

51

Antwoord:

Antwoord: 56 is het zesde element van de NumPy-lijst `y`. Het element 51 van NumPy-lijst `x` heeft dezelfde index.

### Voorbeeld 2.4: een element verwijderen
Je kunt via de *index* elementen uit de NumPy-lijst 'rij' verwijderen met de functie `delete()`.

In [13]:
print(rij)

[ 2  7 10  5]


In [14]:
kortererij = np.delete(rij, 1)   # maak nieuwe lijst die dezelfde is als eerste lijst maar zonder element met index 1
print(kortererij)

[ 2 10  5]


In [15]:
kortsterij = np.delete(rij, [1, 3]) # maak nieuwe lijst die dezelfde is als eerste lijst maar zonder elementen met index 1 en 3
print(kortsterij)

[ 2 10]


### Opdracht 2.4
Beschouw de NumPy-lijst waarnaar wordt verwezen met de variabele `rij2`. Maak hiervan een kortere NumPy-lijst door de elementen 8, 12 en 18 te verwijderen m.b.v. `delete()`. 

In [16]:
# voorbeeldoplossing
kortererij2 = np.delete(rij2, [3, 5, 8])
print(kortererij2)

[ 2  4  6 10 14 16 20]


### Voorbeeld 2.5: NumPy-lijsten samenvoegen
Je kunt NumPy-lijsten samenvoegen met de functie `append()`.

In [17]:
samengevoegdelijst = np.append(rij, rij2)
print(samengevoegdelijst)

[ 2  7 10  5  2  4  6  8 10 12 14 16 18 20]


### Opdracht 2.5
Voeg de NumPy-lijst `x` samen met de NumPy-lijst `rij`. `x` komt eerst in de samengestelde NumPy-lijst.

In [18]:
# voorbeeldoplossing
samengevoegdelijst2 = np.append(x, rij)
print(samengevoegdelijst2)

[ 28  40  41  44  45  51  54  57  57  60  62  72  74  90 113   2   7  10
   5]


### Voorbeeld 2.6: variantie en standaardafwijking
Met de functies `var()` en `std()` vraag je de variantie en de standaardafwijking op van alle elementen uit een NumPy-lijst.

In [19]:
x_variantie = np.var(x)
x_standaardafwijking = np.std(x)
print(x_variantie, x_standaardafwijking)

429.62666666666667 20.727437532571813


### Opdracht 2.6
- Bereken de variantie van de elementen van de NumPy-lijst `y`.
- Bereken de standaardafwijking van de elementen van de NumPy-lijst `y`. 

In [20]:
# voorbeeldoplossing
y_variantie = np.var(y)
y_standaardafwijking = np.std(y)

### Opdracht 2.6
Beschouw de NumPy-lijst waarnaar wordt verwezen met de variabele `rij2`. Maak hiervan een kortere NumPy-lijst door de elementen 8, 12 en 18 te verwijderen m.b.v. `delete()`. 

### Voorbeeld 2.7:  de lengte van een NumPy-lijst
Je kunt met de functie `len()` het aantal elementen van een NumPy-lijst opvragen.

In [21]:
len(rij)

4

### Opdracht 2.7
Vraag de lengte op van de NumPy-lijst `x`. 

In [22]:
# voorbeeldoplossing
len(x)

15

### Voorbeeld 2.8: een deel van een NumPy-lijst selecteren
Het viel je misschien al op in voorbeeld 2.2 dat je a.d.h.v. de *index* een deel van een NumPy-lijst kunt selecteren.

In [23]:
stukvanrij2 = rij2[3: 7]      # halfopen interval [3, 7[
print(stukvanrij2)

[ 8 10 12 14]


In [24]:
eerstehelftrij2 = rij2[:5]     
tweedehelftrij2 = rij2[5:]
print(eerstehelftrij2)
print(tweedehelftrij2)

[ 2  4  6  8 10]
[12 14 16 18 20]


### Opdracht 2.8
- Beschouw de NumPy-lijst waarnaar wordt verwezen met de variabele `y`. Maak een nieuwe NumPy-lijst die bestaat uit de elementen met index 1 t.e.m. index 8. Verwijs naar de nieuwe NumPy-lijst met `y_nieuw`. 
- Maak een nieuwe NumPy-lijst `nieuwelijst` die de drie laatste elementen van `rij` bevat.
- Maak een nieuwe NumPy-lijst `nieuwstelijst` die eerste drie elementen van `x` bevat.

In [25]:
# voorbeeldoplossing
y_nieuw = y[1:9]
nieuwelijst = rij[1:]
nieuwstelijst = x[:3]
print(y_nieuw)
print(nieuwelijst)
print(nieuwstelijst)

[35 47 35 55 56 50 50 70]
[ 7 10  5]
[28 40 41]


<img src="images/cclic.png" alt="Banner" align="left" width="100"/><br><br>
Notebook Python in wiskunde, zie Computationeel denken - Programmeren in Python van <a href="http://www.aiopschool.be">AI Op School</a>, van F. wyffels & N. Gesquière in licentie gegeven volgens een <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Naamsvermelding-NietCommercieel-GelijkDelen 4.0 Internationaal-licentie</a>. 